Update unikernel with new build from builds.robur.coop

PizieDust opened the pull request:
This PR is a continuation of the work started in https://github.com/robur-coop/mollymawk/pull/102 and is related to #54 

In this PR, we :
- [x] download the binary from builds.robur.coop,
- [x] perform a manifest check to ensure that there's no changes in the bridge or block devices.
- [x] check that a binary exists from the latest build before attempting to download it
- [ ] check if the vm should be compressed or not
- [x] modify or update the `args` passed in the config to albatross
if everything checks out, then we pass this binary to albatross to perform the update.

reynir commented on unikernel.ml:
```diff
@@ -1240,6 +1241,112 @@ struct
+            | Error (`Msg err) ->
+                Logs.err (fun m ->
+                    m
+                      "builds.robur.coop: Error while fetching the binary of \
+                       with error: %s"
```
This error message is missing some words it seems: "the binary of with error:"

reynir commented on unikernel.ml:
```diff
@@ -1240,6 +1241,112 @@ struct
+                         builds.robur.coop with error " ^ err))
+                  `Internal_server_error
+            | Ok image -> (
+                Logs.info (fun m -> m "The binary is %s" image);
```
I think this would log a binary image!! Too big :D

PizieDust replied:
This slipped into the commit by error. Will remove it. Thank you for catching it.

reynir commented on albatross.ml:
```diff
@@ -91,6 +95,38 @@ struct
+        (function Solo5_elftool.Dev_block_basic name -> Some name | _ -> None)
+        mft.Solo5_elftool.entries
+    in
+    let bridges_diff =
+      String_set.(diff (of_list bridges') (of_list bridges) |> elements)
+    in
+    let blocks_diff =
+      String_set.(
+        diff (of_list block_devices') (of_list block_devices) |> elements)
+    in
+    match (bridges_diff, blocks_diff) with
+    | [], [] -> Ok ()
+    | [], blocks ->
+        Error (`Msg ("Block devices mismatch: " ^ String.concat "," blocks))
+    | bridges, [] ->
+        Error (`Msg ("Network devices mismatch: " ^ String.concat "," bridges))
+    | bridges, blocks ->
+        Error
+          (`Msg
+             ("Block devices mismatch: " ^ String.concat "," blocks
+            ^ " and network devices mismatch: " ^ String.concat "," bridges))
+
```
So you could check the set difference both ways is empty:

```OCaml
let req_only_bridges = String_set.diff bridges bridges'
and mft_only_bridges = String_set.diff bridges' bridges in
...
match (req_only_bridges, mft_only_bridges, req_only_blocks, mft_only_blocls) with
| [], [], [], [] -> Ok ()
| req_only_bridges, mft_only_bridges, ... -> Error (`Msg "Network devices only in manifest: "^String.concat mft_only_bridges ^"...
```

reynir commented on albatross.ml:
```diff
@@ -97,35 +97,91 @@ struct
+             ^ String.concat "," mft_only_bridges))
+    | [], [], req_only_blocks, [] ->
+        Error
+          (`Msg
+             ("Block devices missing from manifest: "
+             ^ String.concat "," req_only_blocks))
+    | [], [], [], mft_only_blocks ->
+        Error
+          (`Msg
+             ("Block devices only in manifest: "
+             ^ String.concat "," mft_only_blocks))
+    | req_only_bridges, req_only_blocks, [], [] ->
+        Error
+          (`Msg
+             ("Network devices missing from manifest: "
+             ^ String.concat "," req_only_bridges
+             ^ " and block devices missing from manifest: "
+             ^ String.concat "," req_only_blocks))
+    | [], req_only_blocks, mft_only_bridges, [] ->
+        Error
+          (`Msg
+             ("Network devices only in manifest: "
+             ^ String.concat "," mft_only_bridges
+             ^ " and block devices missing from manifest: "
+             ^ String.concat "," req_only_blocks))
+    | [], [], mft_only_blocks, req_only_bridges ->
+        Error
+          (`Msg
+             ("Network devices missing from manifest: "
+             ^ String.concat "," req_only_bridges
+             ^ " and block devices only in manifest: "
+             ^ String.concat "," mft_only_blocks))
+    | req_only_bridges, [], mft_only_blocks, [] ->
+        Error
+          (`Msg
+             ("Network devices missing from manifest: "
+             ^ String.concat "," req_only_bridges
+             ^ " and block devices only in manifest: "
+             ^ String.concat "," mft_only_blocks))
+    | req_only_bridges, req_only_blocks, mft_only_bridges, mft_only_blocks ->
```
I think this could be refactored (e.g. "_ devices missing from manifest", "_ devices only in manifest"), that is, factor out the device type.

PizieDust replied:
Yes I'll send in a commit with better/shorter messages

reynir reviewed (approved):
I think this looks fine. I have not tried to review the javascript changes, and I may be missing context.

reynir commented on albatross.ml:
```diff
@@ -91,6 +95,94 @@ struct
             }
     | Error (`Msg err) -> Error err
 
+  let manifest_devices_match ~bridges ~block_devices binary =
+    let* mft : Solo5_elftool.mft = Solo5_elftool.query_manifest binary in
```
Here we (unfortunately) need to convert the binary to a `Owee_buf.t` (bigarray):
```suggestion
  let manifest_devices_match ~bridges ~block_devices binary =
    let b = Bigarray.Array1.create Bigarray.Int8_unsigned Bigarray.c_layout (String.length b) in
    for i = 0 to String.length b - 1 do
      buf.{i} <- String.get_uint8 b i
    done;
    let binary = b in
    let* mft : Solo5_elftool.mft = Solo5_elftool.query_manifest binary in
```
Otherwise we get a type error. With this PR we should be able to do it without copying too much: https://git.robur.coop/robur/ocaml-solo5-elftool/pulls/2
So while this is very inefficient I'll submit a fix to that Soon™.

reynir commented:
Sorry there were still some C stubs in the owee fork that broke things. It's fixed in https://github.com/robur-coop/owee/commit/b1f94753a6a873a21103cb0d19a501b77cdf28d2

PizieDust commented:
@hannesm Please feel free to merge :)

hannesm commented:
Thanks for your reminder, @PizieDust. I didn't look at this change at all. If you'd like to have this PR reviewed by me, it will need to wait some more time.

OTOH, Reynir reviewed it, so it should be good. I'd appreciate if we avoid a pin in `config.ml` and wait for the solo5-elftool 0.4 to be released (which, if I understand Reynir, will happen soon)!?

PizieDust commented:
> Thanks for your reminder, @PizieDust. I didn't look at this change at all. If you'd like to have this PR reviewed by me, it will need to wait some more time.
> OTOH, Reynir reviewed it, so it should be good. I'd appreciate if we avoid a pin in `config.ml` and wait for the solo5-elftool 0.4 to be released (which, if I understand Reynir, will happen soon)!?

Yes we shouldn't merge with the pin. We'll wait for the release.

hannesm commented on albatross.ml:
```diff
@@ -91,6 +95,98 @@ struct
+    for i = 0 to String.length binary - 1 do
+      b.{i} <- String.get_uint8 binary i
+    done;
+    let binary = b in
```
can we name the binding `b` above just `binary` and then don't need this alias?

reynir commented:
This will need more changes for ocaml-solo5-elftool then. I'll work on that
and then merge (I didn't yet look at the new error message code).

tirs. 4. feb. 2025 16.46 skrev Hannes Mehnert ***@***.***>:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In albatross.ml
> <https://github.com/robur-coop/mollymawk/pull/105#discussion_r1941427449>:
>
> > @@ -91,6 +95,98 @@ struct
>              }
>      | Error (`Msg err) -> Error err
>
> +  let manifest_devices_match ~bridges ~block_devices binary =
> +    let b =
> +      Bigarray.Array1.create Bigarray.Int8_unsigned Bigarray.c_layout
> +        (String.length binary)
> +    in
> +    for i = 0 to String.length binary - 1 do
> +      b.{i} <- String.get_uint8 binary i
> +    done;
> +    let binary = b in
>
> can we name the binding b above just binary and then don't need this
> alias?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/robur-coop/mollymawk/pull/105#pullrequestreview-2593286106>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACVC324MDX7VDZ6GWX3XBT2ODOETAVCNFSM6AAAAABWCZG5HWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKOJTGI4DMMJQGY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

hannesm commented on albatross.ml:
```diff
@@ -91,6 +95,98 @@ struct
+             ^ " devices missing from manifest and "
+             ^ String.concat "," mft_only_bridges
+             ^ String.concat "," mft_only_blocks
+             ^ " devices only in manifest."))
```
you refer to `manifest` in this error message a lot of times, but is this really the wording an operator of mollymawk will understand?

As I understand it, we have the unikernel image and in there we have required devices (what you name manifest); and then we have the command line parameters. So these two sets (actually 4, since it is two for bridges and two for block devices) may disjoint, and if they are the unikernel won't start.

What would be a message an operator understands? Maybe
"You didn't specify the network bridge 'management', which is required by the unikernel."

And
"You specified the network bridge 'private' as command line parameter, which the unikernel does not refer to at all."

Now, it may help to suggest what the operator needs to do:
"Please add a command-line parameter for the 'management' bridge: '--net:private=<my-bridge>'".

"Please remove the command-line parameter '--net=private'"

PizieDust replied:
I totally agree with this. I have a question, for a missing bridge or block device, does the user add it in the block_devices/network_devices section this json, or in the `arguments`. I am asking based on the example you have given which uses command-line parameters. At the moment the only way the user can modify the arguments is by modifying a json such as below.

```json
{
  "name": "some name",
  "typ": "solo5",
  "fail_behaviour": {
    "restart": null,
    "exit_code": [],
    "all_exit_codes": true
  },
  "cpuid": 16,
  "memory": 1024,
  "block_devices": [
    { "name": "data", "host_device": "service", "sector_size": 512 }
  ],
  "network_interfaces": [
    {
      "name": "service",
      "host_device": "internal",
    }
  ],
  "arguments": [
    "--ipv4=192.168.20.11/00", 
    "--ignore-local-git"
  ],
}

```

reynir replied:
the `arguments` field is the boot arguments to the unikernel. The missing bridge would go into the `network_interfaces` list. Similarly, missing block devices go into the `block_devices` list.

hannesm commented on albatross_json.ml:
```diff
@@ -298,3 +298,63 @@ let config_of_json str =
+    match unikernel.argv with
+    | Some args ->
+        [ ("arguments", `List (List.map (fun arg -> `String arg) args)) ]
+    | None -> [])
```
there's in Albatross_json right at the top (line 3) a `unikernel_info` function -- how are they related?

hannesm commented on unikernel.ml:
```diff
@@ -1240,6 +1241,122 @@ struct
+        with
+        | Error (`Msg err) ->
+            Middleware.http_response reqd ~title:"Error"
+              ~data:(`String ("Manifest mismatch: " ^ err))
```
as outlined above, I wouldn't introduce the `Manifest` teminology.

hannesm commented on unikernel.ml:
```diff
@@ -1240,6 +1241,122 @@ struct
+                Logs.err (fun m -> m "Error querying albatross: %s" msg);
+                Middleware.http_response reqd ~title:"Error"
+                  ~data:(`String ("Error querying albatross: " ^ msg))
+                  `Internal_server_error
```
this error could be a bit more precise -- we're 'force_create' a unikernel here.

hannesm commented on unikernel.ml:
```diff
@@ -1240,6 +1241,122 @@ struct
+                      ~data:
+                        (`String
+                           (unikernel_name
+                          ^ " has been updated to the latest build."))
```
maybe to build <ID> where ID could be the sha256 or the build UUID so that the operator could look it up (since there may be while updating an even newer build be produced and uploaded to builds.robur.coop).

hannesm commented on unikernel.ml:
```diff
@@ -1240,6 +1241,122 @@ struct
+                      ~data:(`String (String.escaped err))
+                      `Internal_server_error))
+        | Ok (Some args) -> (
+            match Albatross_json.config_of_json args with
```
I don't quite understand the argument handling:
```OCaml
match Utils.Json.get "unikernel_arguments" json_dict with
    |  arguments -> match Utils.Json.string_or_none "unikernel_arguments" arguments with
      | OK (Some args) -> match Albatross_json.config_of_json args with
```

so in which way need they to be present? json encoded that is encoded as string? why not just as a dict directly in the user parameters?

PizieDust replied:
This is done in the current commit

reynir commented:
The code is now updated to ocaml-solo5-elftool.0.4.0 (pending release). With a temporary pin (now removed) it passes CI \o/

PizieDust commented:
> The code is now updated to ocaml-solo5-elftool.0.4.0 (pending release). With a temporary pin (now removed) it passes CI \o/

Awesome. 
Thank you Reynir

hannesm commented:
This is fine. I don't quite understand why it triggers the opam-monorepo failure, though.

I'll merge to clean up the queue, and will propose the step for albatross 2.4.0
